Book.java

```diff
@@ -0,0 +1,315 @@
+        addAccount(cashAccount, null);
+        addAccount(equityAccount, null);
+    }
+
+
+    private Agent owner;
+    private HashSet<Account> accounts;
+    private HashSet<Account> assetAccounts;
+    private HashSet<Account> liabilityAccounts;
+    private HashSet<Account> equityAccounts;
+    private HashMap<Class<? extends contracts.Contract>, Account> assetAccountsMap;
+    private HashMap<Class<? extends contracts.Contract>, Account> liabilityAccountsMap;
+    private Account cashAccount;
+    private Account equityAccount;
+
+    public double getAssetValue() {
+        double assetTotal = 0;
+        for (Account assetAccount : assetAccounts) {
+            assetTotal+=assetAccount.getBalance();
+        }
+        return assetTotal;
+    }
+
+    public double getLiabilityValue() {
+        double liabilityTotal = 0;
+        for (Account liabilityAccount : liabilityAccounts) {
+            liabilityTotal+=liabilityAccount.getBalance();
+        }
+        return liabilityTotal;
+    }
+
+    public double getEquityValue() {
+        double equityTotal = 0;
+        for (Account equityAccount : equityAccounts) {
+            equityTotal += equityAccount.getBalance();
+        }
+        return equityTotal;
+    }
+
+    public double getAssetValueOf(Class<? extends Contract> contractType) {
+        return assetAccountsMap.get(contractType).getBalance();
+    }
+
+    public double getLiabilityValueOf(Class<? extends Contract> contractType) {
+        return liabilityAccountsMap.get(contractType).getBalance();
+    }
+
+    public double getCash() {
+        return cashAccount.getBalance();
+    }
+
+
+    private void addAccount(Account account, Class<? extends Contract> contractType) {
+        accounts.add(account);
+        switch(account.getAccountType()) {
+
+            case ASSET:
+                assetAccounts.add(account);
+                assetAccountsMap.put(contractType, account);
+                break;
+
+            case LIABILITY:
+                liabilityAccounts.add(account);
+                liabilityAccountsMap.put(contractType, account);
+                break;
+
+            case EQUITY:
+                equityAccounts.add(account);
+
+            // Not sure what to do with INCOME, EXPENSES
+        }
+
+    }
+
+    /**
+     * Adding an asset means debiting the account relevant to that type of contract
+     * and crediting equity.
+     * @param contract an Asset contract to add
+     */
+    public void addAsset(Contract contract) {
+        Account assetAccount = assetAccountsMap.get(contract.getClass());
+
+        if (assetAccount==null) {
+            // If there doesn't exist an Account to hold this type of contract, we create it
+            assetAccount = new Account(contract.getClass().getName(), AccountType.ASSET);
+            addAccount(assetAccount, contract.getClass());
+        }
+
+        // (dr asset, cr equity)
+        assetAccount.debit(contract.getValue());
```

Comment:
I'm not sure if valuation shouldn't be external to the contract, e.g. by means of a function that accepts a contract and returns a value. If it is internal to the contract, then it will be difficult to have different types of valuation. Also, if valuation is done in a more functional way, it will become easier to plug together more complex types of valuation. But that's probably a separate discussion.